docs(SDB): add migration doc MTA-4759 (#3374)
scaleway/docs-content: serverless/sql-databases/api-cli/import-data-to-serverless-sql-databases.mdx

---
meta:
  title: Importing data into Serverless SQL Databases
  description: This page explains how to import your data into Serverless SQL Databases from another database or backup files
content:
  h1: Importing data into Serverless SQL Database
  paragraph: This page explains how to import your data into Serverless SQL Databases from another database or backup files
dates:
  validation: 2024-06-24
  posted: 2024-06-24
---

Serverless SQL Databases offers near-complete PostgreSQL compatibility, and allows you to import your data from any PostgreSQL-compatible database.

<Message type="important">

> The methods listed below are best suited for database sizes below 100 GB (approx.). Above this size, we recommend you use dedicated tooling to perform this operation, as your local storage size and bandwidth will create bottlenecks, and the operation may last several hours.

Reviewer: ```suggestion
The methods listed below are best suited for database sizes below 100 GB (approx.). Above this size, we recommend you use dedicated tooling to import your data, as your local storage size and bandwidth will create bottlenecks, and the operation may last several hours.
```

</Message>

<Macro id="requirements" />

- A Scaleway account logged into the [console](https://console.scaleway.com)
- [Owner](/identity-and-access-management/iam/concepts/#owner) status or [IAM permissions](/identity-and-access-management/iam/concepts/#permission) allowing you to perform actions in the intended Organization

- [Created a target Serverless SQL Database](/serverless/sql-databases/how-to/create-a-database) to import your data into.

> - Checked that the target database you are importing into is empty, or contains data you can overwrite.

Reviewer: ```suggestion
- Checked that the target database you are importing data into is empty, or contains data you can overwrite.
```

The import procedure depends on your data source:
    - [Importing data from an existing PostgreSQL database](#importing-data-from-an-existing-postgresql-database)
      - [Using pg_dump](#using-pgdump)
      - [Using pgAdmin](#using-pgadmin)
    - [Importing data from an existing PostgreSQL backup](#importing-data-from-an-existing-postgresql-backup)
    - [Importing data from a file (.csv, .txt, etc.)](#importing-data-from-a-file-csv-txt-etc)

## Importing data from an existing PostgreSQL database

### Using pg_dump

#### Prerequisites

> To complete this procedure, you must have:
> 
> - Installed PostgreSQL 16 (or newer) with [pg_dump](https://www.postgresql.org/docs/current/app-pgdump.html) and [pg_restore](https://www.postgresql.org/docs/current/app-pgrestore.html) (bundled with the default PostgreSQL installation).

Reviewer: ```suggestion
To complete this procedure, you must have installed PostgreSQL 16 (or newer) with [pg_dump](https://www.postgresql.org/docs/current/app-pgdump.html) and [pg_restore](https://www.postgresql.org/docs/current/app-pgrestore.html) (bundled with the default PostgreSQL installation).
```

### Using pgAdmin

#### Prerequisites

To complete this procedure, you must have:

- Installed [pg_Admin](https://www.pgadmin.org/).

- Ensured your PGAdmin version is compatible with the PostgreSQL version of the database you will import your data into. Currently, pgAdmin 7.8 or newer is required to fully work with PostgreSQL 16.

#### Downloading the backup on your local machine

1. Open pgAdmin.

2. Connect to your existing database server by selecting it in the left-hand menu.

3. In the **Databases** submenu, connect to the database you want to export by selecting it.

4. In the left-hand menu, right-click the database you want to export, and select **Backup**.

5. Enter a name for the backup file (e.g. `my-backup`), and keep the default **format** and **compression** options.

6. Click **Backup**. The backup creation process starts, and you can follow its progress by clicking **View processes**.

7. When finished, make sure your backup is downloaded on your local machine. By default, the backup location is `Users/{username}/my-backup` on MacOS.

#### Loading your data in your Serverless SQL Database

> 1. Register your new database in pgAdmin. Refer to the [How to connect to a Serverless SQL Database](/serverless/sql-databases/how-to/connect-to-a-database/#how-to-set-up-credentials) documentation for more information.

Reviewer: ```suggestion
1. Register your new database in pgAdmin. Refer to the [How to connect to a Serverless SQL Database](/serverless/sql-databases/how-to/connect-to-a-database/#how-to-set-up-credentials) documentation page for more information.
```

2. Connect to your database server by selecting it in the left-hand menu.

3. Select your database in the **Databases* submenu.

4. In the left-hand menu, right-click the name of your new database, then select **Restore**.

5. Enter the file name of your backup (e.g. `my-backup`), and keep the default options.

6. Access the **Data options** tab, and enable the **Do not save - Owner** and **Do not save - Privileges** options.

7. (Optional) If your new database is not empty, access the **Query Options** tab, and enable the **Clean before restore** option. 

    <Message type="tip">

>     This option ensures your database is emptied before loading data to avoid potential conflicts with existing data (such as table names or primary key constraints).

Reviewer: ```suggestion
    This option ensures that your database is emptied before loading data to avoid potential conflicts with existing data (such as table names or primary key constraints).
```

    </Message>

8. Click **Restore**. The restore process starts, and you can follow its progress by clicking **View processes**.

9. When finished, make sure your data is stored in your new database by [connecting to it](/serverless/sql-databases/how-to/connect-to-a-database/), and performing a query.

> If the process fails and some data were partly already transferred, we suggest you activate the **Clean before restore** option in the **Data options** tab to remove partly transferred data.

Reviewer: ```suggestion
If the process fails and some data was already partly transferred, we suggest that you activate the **Clean before restore** option in the **Data options** tab to remove partly transferred data.
```

    ```
    pg_restore --no-privileges --no-owner --clean --if-exists -U {username} -h {host} --port 5432 -d {databasename} my-backup
    ```

    <Message type="note">
    You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM secret Key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 
    </Message>

2. When finished, make sure your data is stored in your new database by [connecting to it](/serverless/sql-databases/how-to/connect-to-a-database/), and performing a query.

## Import Data from a file (.csv, .txt, etc.)

### Using psql

#### Prerequisites

To complete this procedure, you must have:

- Installed PostgreSQL 16 (or newer) with [pg_restore](https://www.postgresql.org/docs/current/app-pgrestore.html) (bundled with the default PostgreSQL installation).

> - A data file corresponding to a single table (named named `my-table.csv` in the following procedure).

Reviewer 2: ```suggestion
- A data file corresponding to a single table (named `my-table.csv` in the following procedure).
```

<Message type="tip">
You can create a `.csv` file from an existing PostgreSQL table with the [COPY TO command](https://www.postgresql.org/docs/current/sql-copy.html), as shown below:

```sql
\COPY {tablename} TO './my-table.csv' WITH DELIMITER ',' CSV;
```
</Message>

> 1. In a termminal, access the folder containing your data file:

Reviewer 2: ```suggestion
1. In a terminal, access the folder containing your data file:
```

---

Serverless SQL Databases offers near-complete PostgreSQL compatibility, and allows you to import your data from any PostgreSQL-compatible database.

<Message type="important">
The methods listed below are best suited for database sizes below 100 GB (approx.). Above this size, we recommend you use dedicated tooling to perform this operation, as your local storage size and bandwidth will create bottlenecks, and the operation may last several hours.
</Message>

The import procedure depends on your data source:
    - [Importing data from an existing PostgreSQL database](#importing-data-from-an-existing-postgresql-database)
      - [Using pg_dump](#using-pgdump)
      - [Using pgAdmin](#using-pgadmin)
      - [Using Airbyte](#using-airbyte)
      - [Using Fivetran](#using-fivetran)
      - [Using Meltano](#using-meltano)
      - [Using Apache NiFi](#using-apache-nifi)
    - [Importing data from an existing PostgreSQL backup](#importing-data-from-an-existing-postgresql-backup)
    - [Importing data from a file (.csv, .txt, etc.)](#importing-data-from-a-file-csv-txt-etc)

## Importing data from an existing PostgreSQL database

### Using pg_dump

#### Before you start

> To complete the actions presented below, you must have:

Reviewer 3: Does it also need the standard requirements macro? (Created a SCW account, IAM rights etc)

    ```
    pg_dump --no-privileges --no-owner -U {username} -h {host} --port {port} -Fc {databasename} > my-backup
    ```

    <Message type="tip">
    You can download specific tables using the `-t` option:

      ```
      pg_dump -t table1name -t table2name --no-privileges --no-owner -U {username} -h {host} --port {port} -Fc {databasename} > my-backup
      ```
    </Message>

2. When finished, make sure the backup is downloaded on your local machine. The default location is in the current directory, under `/my-backup`.

3. Run the command below to import data into your Serverless SQL Database using `pg_restore`. Make sure to replace the placeholders with your Serverless SQL Database connection parameters:

    ```
    pg_restore --no-privileges --no-owner --clean --if-exists -U {username} -h {host} --port 5432 -d {databasename} my-backup
    ```

    <Message type="note">
    You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database.
    </Message>

4. Enter the database password when prompted. 

    <Message type="note">

>     The password is the [IAM Secret Key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 

Reviewer 3: ```suggestion
    The password is the [IAM secret key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 
```

- [Created a target Serverless SQL Database](/serverless/sql-databases/how-to/create-a-database) to import your data into.

- Checked that the target database you are importing into is empty, or contains data you can overwrite.

<Message type="note">
To avoid any statement timeout errors with Managed Database for PostgreSQL (e.g. `ERROR: canceling statement due to statement timeout`), we recommend you set up the PostgreSQL parameter `statement_timeout` to `1200000` (20 minutes), as recommended by Airbyte for other PostgreSQL-compatible data sources. You can modify this parameter in your [database overview in the Scaleway console](https://console.scaleway.com/rdb/instances), by clicking **More**, then **Advanced Settings**.
</Message>

#### Adding your existing database as a source

1. From the Airbyte user interface, click **Sources**.

2. Select **Postgres** from the list of sources.

3. Fill in the required information to connect to your existing database:
    - **Host**
    - **Port**
    - **Database Name**
    - **Username**
    - **Password**

    <Message type="note">

>     You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM Secret Key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 

Reviewer 3: ```suggestion
    You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM secret key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 
```

    </Message>

4. Select an SSL Mode and an SSH Tunnel Method.

5. Select **Detect Changes with Xmin System Column** as **Update Method** (other methods are not supported).

4. Click **Set up source**.

#### Adding your new Serverless SQL Database as a destination

1. Click **Create a connection**.

2. Select **Postgres** from the list of destinations.

3. Fill in the required information to connect to your existing database:
    - **Host**
    - **Port**
    - **Database Name**
    - **Username**
    - **Password**

    <Message type="note">

>     You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM Secret Key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 

Reviewer 3: ```suggestion
    You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM secret key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 
```

<Message type="important">
This procedure does not currently work with Serverless SQL Databases. However, you can import your backup using [Managed Database for PostgreSQL and MySQL](/managed-databases/postgresql-and-mysql/).
</Message>

### Using Fivetran

#### Before you start

To complete the actions presented below, you must have:

- A Fivetran account and can connect to the [Fivetran console](https://fivetran.com/dashboard)

- [Created a target Serverless SQL Database](/serverless/sql-databases/how-to/create-a-database) to import your data into.

- Checked that the target database you are importing into is empty, or contains data you can overwrite.

#### Connecting Fivetran to your Serverless SQL Database

1. From the Fivetran user interface, click **Add destination** in the [destinations list](https://fivetran.com/dashboard/destinations).

2. Enter a name for the destination (e.g. `my-new-database`), then click **Add**.

3. Select **PostgreSQL**, then click **Select**.

4. Fill in the required information to connect to your existing database:

    - **Host**
    - **Port**
    - **Database Name**
    - **Username**
    - **Password**

    <Message type="note">

>     You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM Secret Key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 

Reviewer 3: ```suggestion
    You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM secret key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 
```

    </Message>

7. Click **Save & Continue**.

8. Select the certificate in the list corresponding to `CN=sdb.{region}.scw.cloud`, check its validity, and confirm your selection by clicking **Confirm**. 

    <Message type="tip">
    You can ensure the certificate is fully valid only if all of the following checks are true:
      - Origin is Let's Encrypt
      - The CN field is `sdb.{region}.scw.cloud`
      - The complete certificate value is the same as the one hosted on `sdb.{region}.scw.cloud`
      - The root certificate is the same as [Let's Encrypt ISRG X1 Root certificate](https://letsencrypt.org/certs/isrgrootx1.pem)
    </Message>

#### Connecting Fivetran to your existing database

1. From the Fivetran user interface, click **Add connector** in the [connectors list](https://fivetran.com/dashboard/connectors).

2. Select your destination (for instance `my-new-database`), then click **Select**.

3. Select **PostgreSQL**, then click **Set up**.

4. Fill in the required information to connect to your existing database:

    - **Host**
    - **Port**
    - **Database Name**
    - **Username**
    - **Password**

    <Message type="note">

>     You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM Secret Key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 

Reviewer 3: ```suggestion
    You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM secret key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 
```

1. In a terminal, run the command below to initialize a Meltano project named `migrate-postgresql`:

    ```sh
    meltano init migrate-postgresql
    ```

2. Run the following command to access the newly created directory that contains the project:

    ```sh
    cd migrate-postgresql
    ```

3. Run the following command to add a PostgreSQL-compatible extractor and loader:

    ```sh
    meltano add extractor tap-postgres
    meltano add loader target-postgres
    ```

#### Configuring the extractor to connect to your existing database

1. Run the command below to configure the connection to your existing database:

    ```sh
    meltano config tap-postgres set --interactive
    ```

    A list of settings displays, with a number associated to each parameter.

2. Fill in the required information below by entering the number associated with each parameter:

    - **Database Name** (5)
    - **Host** (13)
    - **Password** (14)
    - **Port** (15)
    - **Username** (32)

3. Select the tables you want to load:

    ```
    meltano select tap-postgres {schema-name}-{table1name}
    meltano select tap-postgres {schema-name}-{table2name}
    ...
    ```

    <Message type="tip">

>     for instance `meltano select tap-postgres public-cities` if you are importing the table `cities` from the schema `public`.

Reviewer 3: ```suggestion
    For instance, select `meltano select tap-postgres public-cities` if you are importing the table `cities` from the schema `public`.
```

    You can check the selected tables and columns with `meltano select tap-postgres --list --all`.
    </Message>

#### Configuring the loader to connect to your new Serverless SQL Database

1. Run the command below to configure the connection to your existing database:

    ```sh
    meltano config tap-postgres set --interactive
    ```

>     A list of settings displays, with a number associated to each parameter.

Reviewer 3: ```suggestion
    A list of settings displays, with a number associated with each parameter.
```

2. Fill in the required information below by entering the number associated with each parameter:

    - **Database Name** (5)
    - **Host** (13)
    - **Password** (14)
    - **Port** (15)
    - **Username** (32)

    <Message type="note">

>     You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM Secret Key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 

Reviewer 3: ```suggestion
    You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM secret key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 
```

    - Database Connection URL: Connection string for your database in the form `jdbc:postgresql://{host}:{port}/{databasename}`
    - Database Driver Class Name: `org.postgresql.Driver`
    - Database Driver Location(s): full path to your PostgreSQL-compatible driver
    - Database User and Password: Credentials to login to your database
    
    <Message type="tip">
    You can check if your connection works by ticking the checkbox, and by clicking **Verify**.
    </Message>

6. Enable your DBCPConnectionPool by clicking on the thunder icon, then on **Enable**.

## Importing data from an existing PostgreSQL backup

### Using pg_restore

#### Before you start

To complete the actions presented below, you must have:

- Installed PostgreSQL 16 (or newer) with [pg_restore](https://www.postgresql.org/docs/current/app-pgrestore.html) (bundled with the default PostgreSQL installation).

- [Created a target Serverless SQL Database](/serverless/sql-databases/how-to/create-a-database) to import your data into.

- Checked that the target database you are importing into is empty, or contains data you can overwrite.

> - a backup file for your database (named `my-backup` in the following procedure). 

Reviewer 3: ```suggestion
- A backup file for your database (named `my-backup` in the following procedure). 
```

### Using psql

#### Before you start

To complete the actions presented below, you must have:

- Installed PostgreSQL 16 (or newer) with [pg_restore](https://www.postgresql.org/docs/current/app-pgrestore.html) (bundled with the default PostgreSQL installation).

- [Created a target Serverless SQL Database](/serverless/sql-databases/how-to/create-a-database) to import your data into.

- Checked that the target database you are importing into is empty, or contains data you can overwrite.

- A data file corresponding to a single table (named named `my-table.csv` in the following procedure).

<Message type="tip">
You can create a `.csv` file from an existing PostgreSQL table with the [COPY TO command](https://www.postgresql.org/docs/current/sql-copy.html), as shown below:

```sql
\COPY {tablename} TO './my-table.csv' WITH DELIMITER ',' CSV;
```
</Message>

1. In a termminal, access the folder containing your data file:

    ```
    cd path/to/my-table.csv
    ```

2. Connect to your Serverless SQL Database using psql:

    ```sh
    psql "postgresql://{username}:{password}@{host}:5432/{databasename}?sslmode=require"
    ```

    <Message type="note">

>     You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM Secret Key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 

Reviewer 3: ```suggestion
    You can find this information on the [overview page](https://console.scaleway.com/serverless-db/databases) of your database. The password is the [IAM secret key](/identity-and-access-management/iam/concepts/#api-key) corresponding to the [IAM user](/identity-and-access-management/iam/concepts/#user) or [IAM application](/identity-and-access-management/iam/concepts/#application) you are connecting with. 
```
